### PR TITLE
fix: correct count(*) alias

### LINKED ISSUE #6447: Cannot have column named "COUNT(*)"
### Describe the bug

After the changes in https://github.com/apache/arrow-datafusion/pull/6010, it's no longer possible to have a column named "COUNT(*)" (even with quotes).  A column with this name is renamed to "COUNT(UInt8(1))".

### To Reproduce

Before #6010:

```
❯ with tbl as (SELECT 0 as "COUNT(*)") SELECT * from tbl;
+----------+
| COUNT(*) |
+----------+
| 0        |
+----------+
```

After #6010

```
❯ with tbl as (SELECT 0 as "COUNT(*)") SELECT * from tbl;
+-----------------+
| COUNT(UInt8(1)) |
+-----------------+
| 0               |
+-----------------+
```

### Expected behavior

I expect it to be possible to have a column named "COUNT(*)" without it being automatically renamed.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jackwener**
# Which issue does this PR close?

Closes #6447.

# Rationale for this change

# What changes are included in this PR?

after replace COUNT(*), we should add a alias to keep name unchanged.

# Are these changes tested?

Yes

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/core/tests/sqllogictests/test_files/subquery.slt [hunk 1]
```diff
@@ -846,65 +842,60 @@ query TT
...
 --Left Join: t1.t1_int = __scalar_sq_1.t2_int
 ----TableScan: t1 projection=[t1_id, t1_int]
 ----SubqueryAlias: __scalar_sq_1
-------Projection: COUNT(UInt8(1)) + Int64(2) AS cnt_plus_2, t2.t2_int, COUNT(UInt8(1)), __always_true
---------Aggregate: groupBy=[[t2.t2_int, Boolean(true) AS __always_true]], aggr=[[COUNT(UInt8(1))]]
-----------TableScan: t2 projection=[t2_int]
+------Projection: COUNT(*) + Int64(2) AS cnt_plus_2, t2.t2_int
+--------Filter: COUNT(*) = Int64(0)
+----------Aggregate: groupBy=[[t2.t2_int]], aggr=[[COUNT(UInt8(1)) AS COUNT(*)]]
+------------TableScan: t2 projection=[t2_int]
```
Comment (jackwener) [inline datafusion/core/tests/sqllogictests/test_files/subquery.slt:852]: cc @mingmwang 

I'm not sure whether new plan is right.

File: datafusion/core/tests/sqllogictests/test_files/subquery.slt [hunk 2]
```diff
@@ -495,9 +495,9 @@ explain SELECT t1_id, t1_name FROM t1 WHERE EXISTS (SELECT count(*) FROM t2 WHER
 logical_plan
 Filter: EXISTS (<subquery>)
 --Subquery:
-----Projection: COUNT(UInt8(1))
```
Comment (jackwener) [inline datafusion/core/tests/sqllogictests/test_files/subquery.slt:498]: This PR change plan in `subquery` shape.

We need notice it carefully.

**Comment by alamb** [comment]
There appears to be a real diff in https://github.com/apache/arrow-datafusion/actions/runs/5656366753/job/15323500341?pr=7081

```
External error: query result mismatch:
[SQL] explain select
            l_returnflag,
            l_linestatus,
            sum(l_quantity) as sum_qty,
            sum(l_extendedprice) as sum_base_price,
            sum(l_extendedprice * (1 - l_discount)) as sum_disc_price,
            sum(l_extendedprice * (1 - l_discount) * (1 + l_tax)) as sum_charge,
            avg(l_quantity) as avg_qty,
            avg(l_extendedprice) as avg_price,
            avg(l_discount) as avg_disc,
            count(*) as count_order
        from
            lineitem
        where
                l_shipdate <= date '1998-09-02'
        group by
            l_returnflag,
            l_linestatus
        order by
            l_returnflag,
            l_linestatus;
[Diff] (-expected|+actual)
    logical_plan
    Sort: lineitem.l_returnflag ASC NULLS LAST, lineitem.l_linestatus ASC NULLS LAST
-   --Projection: lineitem.l_returnflag, lineitem.l_linestatus, SUM(lineitem.l_quantity) AS sum_qty, SUM(lineitem.l_extendedprice) AS sum_base_price, SUM(lineitem.l_extendedprice * Int64(1) - lineitem.l_discount) AS sum_disc_price, SUM(lineitem.l_extendedprice * Int64(1) - lineitem.l_discount * Int64(1) + lineitem.l_tax) AS sum_charge, AVG(lineitem.l_quantity) AS avg_qty, AVG(lineitem.l_extendedprice) AS avg_price, AVG(lineitem.l_discount) AS avg_disc, COUNT(UInt8(1)) AS count_order
-   ----Aggregate: groupBy=[[lineitem.l_returnflag, lineitem.l_linestatus]], aggr=[[SUM(lineitem.l_quantity), SUM(lineitem.l_extendedprice), SUM(lineitem.l_extendedprice * (Decimal128(Some(1),20,0) - lineitem.l_discount)Decimal128(Some(1),20,0) - lineitem.l_discountlineitem.l_discountDecimal128(Some(1),20,0)lineitem.l_extendedprice AS lineitem.l_extendedprice * Decimal128(Some(1),20,0) - lineitem.l_discount) AS SUM(lineitem.l_extendedprice * Int64(1) - lineitem.l_discount), SUM(lineitem.l_extendedprice * (Decimal128(Some(1),20,0) - lineitem.l_discount)Decimal128(Some(1),20,0) - lineitem.l_discountlineitem.l_discountDecimal128(Some(1),20,0)lineitem.l_extendedprice AS lineitem.l_extendedprice * Decimal128(Some(1),20,0) - lineitem.l_discount * (Decimal128(Some(1),20,0) + lineitem.l_tax)) AS SUM(lineitem.l_extendedprice * Int64(1) - lineitem.l_discount * Int64(1) + lineitem.l_tax), AVG(lineitem.l_quantity), AVG(lineitem.l_extendedprice), AVG(lineitem.l_discount), COUNT(UInt8(1))]]
+   --Projection: lineitem.l_returnflag, lineitem.l_linestatus, SUM(lineitem.l_quantity) AS sum_qty, SUM(lineitem.l_extendedprice) AS sum_base_price, SUM(lineitem.l_extendedprice * Int64(1) - lineitem.l_discount) AS sum_disc_price, SUM(lineitem.l_extendedprice * Int64(1) - lineitem.l_discount * Int64(1) + lineitem.l_tax) AS sum_charge, AVG(lineitem.l_quantity) AS avg_qty, AVG(lineitem.l_extendedprice) AS avg_price, AVG(lineitem.l_discount) AS avg_disc, COUNT(*) AS count_order
+   ----Aggregate: groupBy=[[lineitem.l_returnflag, lineitem.l_linestatus]], aggr=[[SUM(lineitem.l_quantity), SUM(lineitem.l_extendedprice), SUM(lineitem.l_extendedprice * (Decimal128(Some(1),20,0) - lineitem.l_discount)Decimal128(Some(1),20,0) - lineitem.l_discountlineitem.l_discountDecimal128(Some(1),20,0)lineitem.l_extendedprice AS lineitem.l_extendedprice * Decimal128(Some(1),20,0) - lineitem.l_discount) AS SUM(lineitem.l_extendedprice * Int64(1) - lineitem.l_discount), SUM(lineitem.l_extendedprice * (Decimal128(Some(1),20,0) - lineitem.l_discount)Decimal128(Some(1),20,0) - lineitem.l_discountlineitem.l_discountDecimal128(Some(1),20,0)lineitem.l_extendedprice AS lineitem.l_extendedprice * Decimal128(Some(1),20,0) - lineitem.l_discount * (Decimal128(Some(1),20,0) + lineitem.l_tax)) AS SUM(lineitem.l_extendedprice * Int64(1) - lineitem.l_discount * Int64(1) + lineitem.l_tax), AVG(lineitem.l_quantity), AVG(lineitem.l_extendedprice), AVG(lineitem.l_discount), COUNT(UInt8(1)) AS COUNT(*)]]
    ------Projection: lineitem.l_extendedprice * (Decimal128(Some(1),20,0) - lineitem.l_discount) AS lineitem.l_extendedprice * (Decimal128(Some(1),20,0) - lineitem.l_discount)Decimal128(Some(1),20,0) - lineitem.l_discountlineitem.l_discountDecimal128(Some(1),20,0)lineitem.l_extendedprice, lineitem.l_quantity, lineitem.l_extendedprice, lineitem.l_discount, lineitem.l_tax, lineitem.l_returnflag, lineitem.l_linestatus
    --------Filter: lineitem.l_shipdate <= Date32("10471")
    ----------TableScan: lineitem projection=[l_quantity, l_extendedprice, l_discount, l_tax, l_returnflag, l_linestatus, l_shipdate], partial_filters=[lineitem.l_shipdate <= Date32("10471")]
    physical_plan
    SortPreservingMergeExec: [l_returnflag@0 ASC NULLS LAST,l_linestatus@1 ASC NULLS LAST]
    --SortExec: expr=[l_returnflag@0 ASC NULLS LAST,l_linestatus@1 ASC NULLS LAST]
-   ----ProjectionExec: expr=[l_returnflag@0 as l_returnflag, l_linestatus@1 as l_linestatus, SUM(lineitem.l_quantity)@2 as sum_qty, SUM(lineitem.l_extendedprice)@3 as sum_base_price, SUM(lineitem.l_extendedprice * Int64(1) - lineitem.l_discount)@4 as sum_disc_price, SUM(lineitem.l_extendedprice * Int64(1) - lineitem.l_discount * Int64(1) + lineitem.l_tax)@5 as sum_charge, AVG(lineitem.l_quantity)@6 as avg_qty, AVG(lineitem.l_extendedprice)@7 as avg_price, AVG(lineitem.l_discount)@8 as avg_disc, COUNT(UInt8(1))@9 as count_order]
-   ------AggregateExec: mode=FinalPartitioned, gby=[l_returnflag@0 as l_returnflag, l_linestatus@1 as l_linestatus], aggr=[SUM(lineitem.l_quantity), SUM(lineitem.l_extendedprice), SUM(lineitem.l_extendedprice * Int64(1) - lineitem.l_discount), SUM(lineitem.l_extendedprice * Int64(1) - lineitem.l_discount * Int64(1) + lineitem.l_tax), AVG(lineitem.l_quantity), AVG(lineitem.l_extendedprice), AVG(lineitem.l_discount), COUNT(UInt8(1))]
+   ----ProjectionExec: expr=[l_returnflag@0 as l_returnflag, l_linestatus@1 as l_linestatus, SUM(lineitem.l_quantity)@2 as sum_qty, SUM(lineitem.l_extendedprice)@3 as sum_base_price, SUM(lineitem.l_extendedprice * Int64(1) - lineitem.l_discount)@4 as sum_disc_price, SUM(lineitem.l_extendedprice * Int64(1) - lineitem.l_discount * Int64(1) + lineitem.l_tax)@5 as sum_charge, AVG(lineitem.l_quantity)@6 as avg_qty, AVG(lineitem.l_extendedprice)@7 as avg_price, AVG(lineitem.l_discount)@8 as avg_disc, COUNT(*)@9 as count_order]
+   ------AggregateExec: mode=FinalPartitioned, gby=[l_returnflag@0 as l_returnflag, l_linestatus@1 as l_linestatus], aggr=[SUM(lineitem.l_quantity), SUM(lineitem.l_extendedprice), SUM(lineitem.l_extendedprice * Int64(1) - lineitem.l_discount), SUM(lineitem.l_extendedprice * Int64(1) - lineitem.l_discount * Int64(1) + lineitem.l_tax), AVG(lineitem.l_quantity), AVG(lineitem.l_extendedprice), AVG(lineitem.l_discount), COUNT(*)]
    --------CoalesceBatchesExec: target_batch_size=8192
    ----------RepartitionExec: partitioning=Hash([l_returnflag@0, l_linestatus@1], 4), input_partitions=4
-   ------------AggregateExec: mode=Partial, gby=[l_returnflag@5 as l_returnflag, l_linestatus@6 as l_linestatus], aggr=[SUM(lineitem.l_quantity), SUM(lineitem.l_extendedprice), SUM(lineitem.l_extendedprice * Int64(1) - lineitem.l_discount), SUM(lineitem.l_extendedprice * Int64(1) - lineitem.l_discount * Int64(1) + lineitem.l_tax), AVG(lineitem.l_quantity), AVG(lineitem.l_extendedprice), AVG(lineitem.l_discount), COUNT(UInt8(1))]
+   ------------AggregateExec: mode=Partial, gby=[l_returnflag@5 as l_returnflag, l_linestatus@6 as l_linestatus], aggr=[SUM(lineitem.l_quantity), SUM(lineitem.l_extendedprice), SUM(lineitem.l_extendedprice * Int64(1) - lineitem.l_discount), SUM(lineitem.l_extendedprice * Int64(1) - lineitem.l_discount * Int64(1) + lineitem.l_tax), AVG(lineitem.l_quantity), AVG(lineitem.l_extendedprice), AVG(lineitem.l_discount), COUNT(*)]
    --------------ProjectionExec: expr=[l_extendedprice@1 * (Some(1),20,0 - l_discount@2) as lineitem.l_extendedprice * (Decimal128(Some(1),20,0) - lineitem.l_discount)Decimal128(Some(1),20,0) - lineitem.l_discountlineitem.l_discountDecimal128(Some(1),20,0)lineitem.l_extendedprice, l_quantity@0 as l_quantity, l_extendedprice@1 as l_extendedprice, l_discount@2 as l_discount, l_tax@3 as l_tax, l_returnflag@4 as l_returnflag, l_linestatus@5 as l_linestatus]
    ----------------CoalesceBatchesExec: target_batch_size=8192
    ------------------FilterExec: l_shipdate@6 <= 10471
    --------------------CsvExec: file_groups={4 groups: [[WORKSPACE_ROOT/datafusion/core/tests/sqllogictests/test_files/tpch/data/lineitem.tbl:0..18561749], [WORKSPACE_ROOT/datafusion/core/tests/sqllogictests/test_files/tpch/data/lineitem.tbl:18561749..37[1234](https://github.com/apache/arrow-datafusion/actions/runs/5656366753/job/15323500341?pr=7081#step:7:1235)98], [WORKSPACE_ROOT/datafusion/core/tests/sqllogictests/test_files/tpch/data/lineitem.tbl:37123498..55685247], [WORKSPACE_ROOT/datafusion/core/tests/sqllogictests/test_files/tpch/data/lineitem.tbl:55685247..74246996]]}, projection=[l_quantity, l_extendedprice, l_discount, l_tax, l_returnflag, l_linestatus, l_shipdate], has_header=false
at tests/sqllogictests/test_files/tpch/./q1.slt.part:19
at tests/sqllogictests/test_files/tpch/tpch.slt:21
```

File: datafusion/core/src/execution/context.rs [hunk 3]
```diff
@@ -2307,11 +2307,11 @@ mod tests {
 
         assert_eq!(results.len(), 1);
         let expected = vec![
-            "+--------------+--------------+-----------------+",
-            "| SUM(test.c1) | SUM(test.c2) | COUNT(UInt8(1)) |",
-            "+--------------+--------------+-----------------+",
-            "| 10           | 110          | 20              |",
-            "+--------------+--------------+-----------------+",
+            "+--------------+--------------+----------+",
```
Comment (alamb) [inline datafusion/core/src/execution/context.rs:2310]: that is certainly much nicer ❤️

File: datafusion/core/tests/sqllogictests/test_files/subquery.slt [hunk 4]
```diff
@@ -714,106 +714,103 @@ query TT
...
+------Projection: COUNT(*), t2.t2_int
+--------Aggregate: groupBy=[[t2.t2_int]], aggr=[[COUNT(UInt8(1)) AS COUNT(*)]]
 ----------TableScan: t2 projection=[t2_int]
 
 query II rowsort
 SELECT t1_id, (SELECT count(*) FROM t2 WHERE t2.t2_int = t1.t1_int) from t1
 ----
 11 1
-22 0
+22 NULL
```
Comment (alamb) [inline datafusion/core/tests/sqllogictests/test_files/subquery.slt:729]: this seems wrong -- `COUNT(*)` should never produce `NULL` values even if it doesn't see any non null inputs


The 
```
CASE WHEN __scalar_sq_1.__always_true IS NULL THEN Int64(0) ELSE __scalar_sq_1.COUNT(UInt8(1)) END AS COUNT(UInt8(1))
``` 

Appears to be doing that translation of NULL --> 0 but it has been lost in this PR

File: datafusion/core/tests/sqllogictests/test_files/subquery.slt (hunk 1 above)
Comment (alamb) [reply datafusion/core/tests/sqllogictests/test_files/subquery.slt:852]: If you are referring to `COUNT(*)` I think it should always be 0

**Review by alamb — COMMENTED** [review]
This looks great @jackwener  except for the subquery plan changes. 

Perhaps this code needs some adjustment: https://github.com/apache/arrow-datafusion/blob/main/datafusion/optimizer/src/decorrelate.rs#L371-L406

**Comment by jackwener** [comment]
> This looks great @jackwener except for the subquery plan changes.

Agree with it.

> Perhaps this code needs some adjustment: https://github.com/apache/arrow-datafusion/blob/main/datafusion/optimizer/src/decorrelate.rs#L371-L406

Thanks !

File: datafusion/core/tests/sqllogictests/test_files/subquery.slt (hunk 4 above)
Comment (jackwener) [reply datafusion/core/tests/sqllogictests/test_files/subquery.slt:729]: I will look these code carefully at weekend, thanks @alamb

**Comment by jackwener** [comment]
cc @jiangzhx

File: datafusion/core/tests/sqllogictests/test_files/subquery.slt
Comment (alamb) [inline datafusion/core/tests/sqllogictests/test_files/subquery.slt:1]: The plans in this file look much better

File: datafusion/optimizer/src/analyzer/count_wildcard_rule.rs [hunk 5]
```diff
@@ -15,23 +15,20 @@
...
-    LogicalPlan, Projection, Sort, Subquery, Window,
+    aggregate_function, expr, lit, window_function, Aggregate, Expr, Filter, LogicalPlan,
+    LogicalPlanBuilder, Projection, Sort, Subquery,
 };
-use std::string::ToString;
 use std::sync::Arc;
 
-use crate::analyzer::AnalyzerRule;
-
-pub const COUNT_STAR: &str = "COUNT(*)";
```
Comment (alamb) [inline datafusion/optimizer/src/analyzer/count_wildcard_rule.rs:33]: 🎉

**Review by alamb — APPROVED** [review]
This looks awesome @jackwener  -- thank you.

I think it is the mark of an excellent engineer when bugs are fixed  by removing code. Very well done 🏆

**Comment by jiangzhx** [comment]
> cc @jiangzhx

Thanks for your help, @jackwener. Great job! I'm your big fan!!

**Comment by alamb** [comment]
> Great job! I'm your big fan!!

Me too!

**Pull request merged**